lldb/tools/lldb-dap/lldb-dap.cpp

```diff
@@ -910,6 +911,183 @@ void request_attach(const llvm::json::Object &request) {
+//       "description": "Start position within `line` to search possible
+//       breakpoint locations in. It is measured in UTF-16 code units and the
+//       client capability `columnsStartAt1` determines whether it is 0- or
+//       1-based. If no column is given, the first position in the start line is
+//       assumed."
+//     },
+//     "endLine": {
+//       "type": "integer",
+//       "description": "End line of range to search possible breakpoint
+//       locations in. If no end line is given, then the end line is assumed to
+//       be the start line."
+//     },
+//     "endColumn": {
+//       "type": "integer",
+//       "description": "End position within `endLine` to search possible
+//       breakpoint locations in. It is measured in UTF-16 code units and the
+//       client capability `columnsStartAt1` determines whether it is 0- or
+//       1-based. If no end column is given, the last position in the end line
+//       is assumed."
+//     }
+//   },
+//   "required": [ "source", "line" ]
+// },
+// "BreakpointLocationsResponse": {
+//   "allOf": [ { "$ref": "#/definitions/Response" }, {
+//     "type": "object",
+//     "description": "Response to `breakpointLocations` request.\nContains
+//     possible locations for source breakpoints.",
+//     "properties": {
+//       "body": {
+//         "type": "object",
+//         "properties": {
+//           "breakpoints": {
+//             "type": "array",
+//             "items": {
+//               "$ref": "#/definitions/BreakpointLocation"
+//             },
+//             "description": "Sorted set of possible breakpoint locations."
+//           }
+//         },
+//         "required": [ "breakpoints" ]
+//       }
+//     },
+//     "required": [ "body" ]
+//   }]
+// },
+// "BreakpointLocation": {
+//   "type": "object",
+//   "description": "Properties of a breakpoint location returned from the
+//   `breakpointLocations` request.",
+//   "properties": {
+//     "line": {
+//       "type": "integer",
+//       "description": "Start line of breakpoint location."
+//     },
+//     "column": {
+//       "type": "integer",
+//       "description": "The start position of a breakpoint location. Position
+//       is measured in UTF-16 code units and the client capability
+//       `columnsStartAt1` determines whether it is 0- or 1-based."
+//     },
+//     "endLine": {
+//       "type": "integer",
+//       "description": "The end line of breakpoint location if the location
+//       covers a range."
+//     },
+//     "endColumn": {
+//       "type": "integer",
+//       "description": "The end position of a breakpoint location (if the
+//       location covers a range). Position is measured in UTF-16 code units and
+//       the client capability `columnsStartAt1` determines whether it is 0- or
+//       1-based."
+//     }
+//   },
+//   "required": [ "line" ]
+// },
+void request_breakpointLocations(const llvm::json::Object &request) {
```

Comment:
Do you know how often VSCode makes this request? Is this something that should be cached once we calculate the values? I'm not sure how often this is called or how expensive it is to look up the compilation unit by path.

Reply:
VSCode makes this request after setting a normal line-breakpoint. In particular, it does not send this request on hovers.

As such, we should be fine as long as the performance of this call is in the same ball park as the performance of the `SetBreakpoint` call itself.

The high-level algorithm in this PR should be fine, given that the `SetBreakpoint` itself also iterates the list in a very similar way (see `LineTable::FindLineEntryIndexByFileIndexImpl`.

The implementation of that algorithm might be sub-optimal. Given that we go through the `SB*` interfaces, we are allocating a lot of `shared_ptr`s afaict. It might make sense to implement a `FindPotentialBreakPointsInRange` method directly in the `LineTable`. There we could iterate and filter the line-table without allocating a `SBLineEntry` instance for each entry in the LineTable.

I would appreciate some guidance on the 2nd point: Should I introduce a `LineTable::FindPotentialBreakPointsInRange`?

CC @JDevlieghere

Reply:
I noticed that I will likely need a very similar, yet slightly different logic for "Go To" / "Jump" support in lldb-dap.

As such, I would defer adding a `FindPotentialBreakpointsInRange` until after also implementing "Go To", so I can hopefully unify both of them into a single, new `FindLineEntriesInSourceRange` function used for both "Go To" and column breakpoints.

My plan of action would be:
1. merge this commit as is
2. implement and merge "Go To". Also using the `SB*` APIs
3. add a `LineEntry::FindLineEntriesInSourceRange` function, plump it through the `SB*` APIs and use it from both (1) and (2)

Please let me know in case you disagree with that plan, @ashgti

Reply:
That sounds reasonable to me